fix(router): fix = not parsed in router segment name

DW8Reaper opened the pull request:
fix router segment name parsing to allow segements to container an unscaped = character. Currently if you have a url like /some-site/folder=/some-file then then middle segment "folder=" will stop parsing at the = sign and register that part of the path as just "folder"

Fixes #21381

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
As outlined in this issue (https://github.com/angular/angular/issues/21381) when you have a url where a path segment contains an = character angular will parse that segment as only the characters before the =. For example **password/de/MDAtMNTk=** the last segment will be parsed in angular as only **MDAtMNTk** without the = character

Issue Number: 21381


## What is the new behavior?
After this change you may include any number of = characters in a segment name and correspondingly define those as route paths in your route configuration. 

## Does this PR introduce a breaking change?

- [X] Yes
- [ ] No

I think this PR could potentially be a breaking change since someone might have specifically handled cases where = was ignored by angular. Suppose they had path segments that were base 64 encoded strings for example **IMz3bg==**, those could possibly end with one or two = characters. It is therefore conceivable that someone might have implemented a workaround where if they received the id as **IMz3bg** they could infer based on the missing bit count that two == characters should be added. In that case if angular now accepted = their resource names could suddenly become **IMz3bg====** 

A counter argument would be that if you had logic to determine the number of missing bits if angular now parsed the = there wouldn't be any missing bits and your logic would not trigger

## Other information

Based off of comments in the issue here (https://github.com/angular/angular/issues/21381#issuecomment-783761338) and (https://github.com/angular/angular/issues/21381#issuecomment-953948114) I looked into [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) and [RFC 1630]() and it would seem that there isn't any specific limitation on using = in a segment name. Consider this excerpt from RFC 3986: 

> For example, the semicolon (";") and equals ("=") reserved characters are often used to delimit parameters and parameter values applicable to that segment. The comma (",") reserved character is often used for similar purposes.  For example, one URI producer might use a segment such as "name;v=1.1" to indicate a reference to version 1.1 of "name", whereas another might use a segment such as "name,1.1" to indicate the same.  Parameter types may be defined by scheme-specific semantics, but in most cases the syntax of a parameter is specific to the implementation of the URI's dereferencing algorithm.

This seems to clearly indicate that you can use = and that is further substantiated by angulars special handling of matric parameters which uses the **name;v=1.1** syntax from this excerpt

DW8Reaper committed:
fix(router): fix = not parsed in router segment name

fix router segment name parsing to allow segements to container an unscaped = character. Currently if you have a url like /some-site/folder=/some-file then then middle segment "folder=" will stop parsing at the = sign and register that part of the path as just "folder"

Fixes #21381

DW8Reaper commented:
I did some testing in a sample app and when I have my routing configured as 
```
const routes: Routes = [
  {path: 'another=test', component: TestComponent}
];
```
then it correctly loads **TestComponent** when I go to http://localhost;4200/another=test where previously it would have tried to got to a route **another**. 

There is one thing that may be unexpected and that is after routing completes the new URL will encode the = so the URL becomes http://localhost:4200/another%3Dtest (that route also works though if you try load it). So now I'm wondering if that result after loading the page should keep the = unencoded it kinda seems like it should any opinion on that?

pkozlowski-opensource commented:
@DW8Reaper sorry for the delay on reviewing / merging this one. It is likely that this change introduces a breaking change so we need to run some internal tests and go over their results.

atscott commented:
Note: This is related to the internal issue [b/234604699](http://b/234604699). These issues are related but actually solve distinctly separate parts of the parsing.

The solution here parses `=` in the segment path but specifically excludes it from matrix params.
The internal issue requires the `=` to be parsed in the matrix param value but does not attempt to parse it elsewhere in the path.

We'll need to evaluate both scenarios and may want to include both in the single solution. The [matrix proposal](https://www.w3.org/DesignIssues/MatrixURIs.html) does not appear to take a stance on the `=` character in parameter values (i.e. `/a/b;myKey=myValueWithAn=InTheMiddle)`.

DW8Reaper commented:
@atscott I'd be happy to update mine if you can provide some details on where the other place was that this was changed on your side. Of course if it is simpler to just update things internally that makes sense to and then I can decline this PR

atscott commented:
@DW8Reaper I think essentially you just need to update `const valueMatch = matchMatrixParamSegments(this.remaining);` to be `const valueMatch = matchSegments(this.remaining);`. We can leave this open and give you the credit for the fix. We're testing the change internally, but that'll take a fair bit of time to roll out safely.

DW8Reaper commented:
Cool I can do that, not sure if you saw my other comment @atscott about it not preserving the = when the page navigation completes instead it the % escapes it. 

Should I see if I can fix that as well to make this more complete?

atscott commented:
> Cool I can do that, not sure if you saw my other comment @atscott about it not preserving the = when the page navigation completes instead it the % escapes it.
> 
> Should I see if I can fix that as well to make this more complete?

I didn't see that. Yea, that seems like something we might want to address at the same time if possible.

DW8Reaper committed:
fixup! fix(router): fix = not parsed in router segment name

DW8Reaper commented:
@atscott sorry that it took a while to update this PR but I have changed it so that if now will keep the = sign after routing in the URL bar of the browser

DW8Reaper commented:
@atscott just following up is there anything else I need to do from my side for this PR?

michaelgriscom commented on packages/router/src/url_tree.ts:
```diff
@@ -459,7 +459,8 @@ function encodeUriString(s: string): string {
       .replace(/%24/g, '$')
-      .replace(/%2C/gi, ',');
+      .replace(/%2C/gi, ',')
+      .replace(/%3D/gi, '=');
```
After some additional discussion with @atscott we think it would be better to leave this change for a separate PR, i.e. this PR will handle parsing `=` in router segments and leave them encoded (the same way `=` are currently handled in query parameters), and in a different PR we can look into leaving them unencoded (sorry for the back and forth!)

DW8Reaper replied:
Man I hope I got this straight I think all an all after you comments I was supposed to remove the URL encoding part then which I did by just reverting that whole second commit.   

I wasn't sure if I was still supposed to remove the second call to matchMatrixParamSegments in the parseParam function from this comment: 
> As alluded to in [#47332 (comment)](https://github.com/angular/angular/pull/47332#issuecomment-1251237403) we've got an internal bug using `=` in matrix parameter values that would be convenient to fix in this PR, all that's needed is to leave this line as calling `matchSegments`
> 
> That will also make it worth renaming `matchMatrixParamSegments` to add more precision, maybe something like `matchMatrixKeySegment`

DW8Reaper replied:
Cool I'll update it then but it'll probably only be in about 2 weeks I'm traveling and don't have my notebook with me

DW

DW8Reaper committed:
revert "fixup! fix(router): fix = not parsed in router segment name"

This reverts commit 2279f4d4620eba083a9832ed096890b69a25ec42.

Reverting that commit based off PR feedback that this change should only affect the parsing of sergments and node encoding of the url

DW8Reaper committed:
fixup! revert "fixup! fix(router): fix = not parsed in router segment name"

DW8Reaper commented on packages/router/test/url_serializer.spec.ts:
```diff
@@ -34,6 +34,36 @@ describe('url serializer', () => {
+    expect(tree.root.children.primary.segments[2].path).toEqual('something');
+    expect(tree.root.children.primary.segments[2].parameterMap.keys).toHaveSize(2);
+    expect(tree.root.children.primary.segments[2].parameterMap.get('query')).toEqual('file=test');
+    expect(tree.root.children.primary.segments[2].parameterMap.get('query2')).toEqual('test2');
```
I added a second matrix parameter **query2** just because I was paranoid I might break something.

michaelgriscom reviewed (approved):
Looks great!

@atscott this looks ready to merge to me

atscott commented:
@michaelgriscom Great, thanks for helping out on the review here! I'll take one last look and run a global presubmit

atscott commented:
[Green TGP](https://fusion2.corp.google.com/presubmit/tap/523131268/OCL:523131268:BASE:523114146:1681157726821:af722dc2/targets)

AndrewKushnir commented:
This PR was merged into the repository by commit ee816e171f23759858e94ebf68447c73735d7f78.
